### ISSUE TITLE
Get rid of unsharding

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#3813]
I'm was looking into `return_initial` earlier today, and ran into another case where it would be really useful not to have the unshard visitor discard so much information.

Right now the unshard visitor merges the results from all the shards together, but this has a few problems:

* If you're doing an ordered range get, it has to look at the largest keys for each shard (call them `k_s`), and then discard all rows with keys larger than `\min_s k_s`.  For certain ordered range gets (`r.table('test').order_by(index: 'id')`, any ordered range get that returns very small batches from all the shards).
  - If we got rid of unsharding, we could keep track rows from all the shards separately, and only fetch more from shards that we needed to.
* Some of the new operations we're adding require us to know a per-shard timestamp, but the result visitor produces an undifferentiated slurry of results from different shards, discarding this information.
  - I'm going to hack around this for now by just adding the information as an optional field of `rget_item_t`, but if we got rid of the unshard visitor it would be much more space efficient not to store a copy of this information for every item from the same shard.
* A cleaner interface -- in my opinion it's nuts that the unshard visitor is constructing environments and evaluating ReQL code.

There are some downsides to getting rid of unsharding too, though:

* Certain queries (e.g. `reduce`) will produce more network traffic without unsharding.
  - We could special-case these queries, but only when we actually don't need per-shard information.
* If we do the optimization I mentioned above for range gets, we'll be keeping data hanging around in memory for open streams, which will increase memory usage for certain workloads (especially workloads with large batch sizes).
  - We could keep only the first few elements of these batches around; this would make the logic more complicated and would still lead to a small number of double-reads, but it would make memory usage basically as good as before and still drastically improve performance for the pathological cases I mentioned above.

**Comment by timmaxw** [comment]
* > It means code from the ql layer leaks all over everything -- in my opinion it's nuts that the unshard visitor is constructing environments and evaluating ReQL code.

    The code that calls `.unshard()` doesn't know that ReQL code is being evaluated, so it doesn't feel like an abstraction break to me.

* > Certain queries (e.g. `reduce`) will produce more network traffic without unsharding.

    Are you sure? `.unshard()` is only called from `cluster_namespace_interface_t`, which runs on the same machine as the parser. (Except for sindex status queries, but those are bullshit anyway.)

* My biggest concern about getting rid of unsharding is that it makes ReQL more tightly coupled to clustering. Right now, clustering sees `.shard()`/`.unshard()` as abstract operations that obey a certain set of properties. For example, `response = store.perform(query)` is equivalent to:
    ```
    q1, q2 = query.shard()
    r1 = store.perform(q1)
    r2

**Comment by timmaxw** [comment]
Oops, I hit "comment" early. The above comment will be edited.

**Comment by mlucy** [comment]
> Will this work correctly even if the table is reconfigured or rebalanced? This sounds like a situation where breaking the abstraction I described above could lead to bugs.

We'll probably have to be more clever, but probably most of the new cleverness will have to happen anyway to make changefeeds work during a rebalance.  (The specific thing I was referencing was supporting `return_initial` for changefeeds on tables, which causes us to have to read from the table in multiple batches and reconcile the reads we do with incoming changes we're receiving from that shard.)

**Comment by mlucy** [comment]
> Are you sure? .unshard() is only called from cluster_namespace_interface_t, which runs on the same machine as the parser. (Except for sindex status queries, but those are bullshit anyway.)

I might be wrong about that; I thought someone told me once that we call the unshard visitor once to unshard all the hash shard results, then send the new values over the network and call the visitor again to unshard all the range shard results.

**Comment by mlucy** [comment]
> The code that calls .unshard() doesn't know that ReQL code is being evaluated, so it doesn't feel like an abstraction break to me.

It bothers me mostly because it means that the unshard visitor has to be able to construct a functioning ReQL environment in order to execute code.

**Comment by timmaxw** [comment]
> I thought someone told me once that we call the unshard visitor once to unshard all the hash shard results

I think we used to do that, but now we treat hash shards and range shards the same at the level we're talking about.

> It bothers me mostly because it means that the unshard visitor has to be able to construct a functioning ReQL environment in order to execute code.

This doesn't bother me because I consider the unshard visitor part of ReQL. Maybe it just boils down to a difference between our coding styles.

**Comment by mlucy** [comment]
Yeah, that might just be a mental difference; I consider the shard/unshard step part of clustering.  (It isn't in the `ql` namespace, nothing in the `ql` namespace knows anything about how it works except for knowing a little bit about regions, it knows a bunch of information about how the shards are laid out even though nothing in the `ql` namespace does, etc. etc.)

**Comment by timmaxw** [comment]
The shard/unshard logic is a weird case because it's the only thing that "knows about" both shards and ReQL queries. The clustering logic knows nothing about ReQL queries and the ReQL logic knows nothing about shards. So it doesn't quite fit in either.
